src/k8s/pkg/k8sd/api/certs_refresh.go

```diff
@@ -0,0 +1,263 @@
+				pkix.Name{
+					CommonName:   csr.commonName,
+					Organization: csr.organization,
+				},
+				2048,
+				nil,
+				csr.hostnames,
+				csr.ips,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to generate CSR for %s: %w", csr.name, err)
+			}
+
+			if _, err = client.CertificatesV1().CertificateSigningRequests().Create(ctx, &certificatesv1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: csr.name,
+				},
+				Spec: certificatesv1.CertificateSigningRequestSpec{
+					Request:    []byte(csrPEM),
+					Usages:     csr.usages,
+					SignerName: csr.signerName,
+				},
+			}, metav1.CreateOptions{}); err != nil {
+				return fmt.Errorf("failed to create CSR for %s: %w", csr.name, err)
+			}
+			watcher, err := client.CertificatesV1().CertificateSigningRequests().Watch(ctx, metav1.SingleObject(metav1.ObjectMeta{Name: csr.name}))
+			if err != nil {
+				return fmt.Errorf("failed to watch CSR %s: %w", csr.name, err)
+			}
+
+			defer watcher.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return nil
+				case evt, ok := <-watcher.ResultChan():
+					if !ok {
+						return fmt.Errorf("watch closed for CSR %s", csr.name)
+					}
+
+					request, ok := evt.Object.(*v1.CertificateSigningRequest)
+					if !ok {
+						return fmt.Errorf("expected a CertificateSigningRequest but received %#v", evt.Object)
+					}
+
+					approved, err := isCertificateSigningRequestApproved(request)
+					if err != nil {
+						return fmt.Errorf("csr is in an invalid state: %w", err)
+					}
+
+					if approved && isCertificateSigningRequestIssued(request) {
+						if _, _, err = pkiutil.LoadCertificate(string(request.Status.Certificate), ""); err != nil {
+							log.WithValues("csr", csr.name).Error(err, "CertificateSigningRequest failed")
+							return fmt.Errorf("CertificateSigningRequest failed: %w", err)
+						}
+						*csr.certificate = string(request.Status.Certificate)
+						*csr.key = string(keyPEM)
+						return nil
+					}
+				}
+			}
+		})
+
+	}
+
+	if err := g.Wait(); err != nil {
+		return response.InternalError(fmt.Errorf("failed to generate worker CSRs: %w", err))
+	}
+
+	if _, err = setup.EnsureWorkerPKI(snap, &certificates); err != nil {
+		return response.InternalError(err)
+	}
+
+	// Kubeconfigs
+	if err := setup.Kubeconfig(filepath.Join(snap.KubernetesConfigDir(), "kubelet.conf"), "127.0.0.1:6443", certificates.CACert, certificates.KubeletClientCert, certificates.KubeletClientKey); err != nil {
+		return response.InternalError(fmt.Errorf("failed to generate kubelet kubeconfig: %w", err))
+	}
+	if err := setup.Kubeconfig(filepath.Join(snap.KubernetesConfigDir(), "proxy.conf"), "127.0.0.1:6443", certificates.CACert, certificates.KubeProxyClientCert, certificates.KubeProxyClientKey); err != nil {
+		return response.InternalError(fmt.Errorf("failed to generate kube-proxy kubeconfig: %w", err))
+	}
+
+	// Restart the services
+	log.Info("Restarting kubelet and kube-proxy")
```

Comment:
this would be more fitting as ` log.V(1).WithValues("service", svcName).Info("Restarting service")` inside the `snap.RestartService(...)` instead